WordPress/gutenberg
# Render html in post titles in visual mode and edit HTML in post title in code view

## What?


Updates the Post Title so that:

- it (visually) _renders_ HTML in Visual Mode
- it renders the _raw_ text in Code Mode (HTML tags are displayed and editable).

**Note**: is is intended that you cannot add/edit HTML in visual mode (i.e. there is not toolbar). This could be added as a followup, but the goal of this PR is an iterative step to ensure users can perceive the _presence_ of HTML in the title.

Closes https://github.com/WordPress/gutenberg/issues/38668

## Why?


[HTML is allowed in post titles in WordPress](https://make.wordpress.org/core/handbook/contribute/design-decisions/#unfiltered-html-for-editors-administrators-multisite).

[A recent consensus in Core](https://core.trac.wordpress.org/ticket/57265) is that HTML should be displayed and visible to the user.

Currently in the Block Editor, there is no way for users to _perceive_ or _edit_ HTML that is in the Post Title when in the editor. This is because the HTML is not rendered at all but it is _persisted_ on save.

This provides a poor experience and leads to titles with _hidden_ HTML 👎 

## How?


Updates the Post to:

- render HTML tags in when in Visual Mode - this allows users to perceieve HTML is present
- render raw text in Code Mode - this allows users to see underlying HTML tags and adjust/remove as necessary.

Note `PostTitle` component remains "as was" with no functionality changes. For the "raw text" version a new component is available called `PostTitleRaw`. This is so as to ensure backwards compat for any consumers of the PostTitle block.

## Testing Instructions





http://gutenberg.run/54718

- New Post
- Switch to "Code View" mode (look for "Options" menu in top right of editor and then under that menu click `Code editor`).
- Paste follow into Post Title field
```
<em>Hello</em> I am <strong><em>emphasis</em></strong> I am <a href="#">anchor</a>
```
- See HTML output in raw form with all tags visible.
- Switch to Visual Mode ((look for "Options" menu in top right of editor and then under that menu click `Visual editor`).
- See HTML (visually) _rendered_ as per the tags in the contents. The idea here is that you should be able to visually perceive that there is HTML present.

Now check that you cannot paste HTML:

- New Post
- Paste the following into Post Title field
```
<em>Hello</em> I am <strong><em>emphasis</em></strong> I am <a href="#">anchor</a>
```
- See no HTML has been rendered.
- Switch to "Code view" mode.
- See no raw HTML in post title field.

### Testing Instructions for Keyboard


## Screenshots or screencast 


https://github.com/WordPress/gutenberg/assets/444434/2058e3d7-5c11-49c3-8f8a-6147e35d6b82
